File: python/purehtml/configs/types/object_config.py
from typing import Any
from typing import Dict
from typing import Optional

from purehtml.configs.configs import Config
from purehtml.configs.extract_params import ExtractParams
from purehtml.configs.selector_match_params import GetSelectorMatchesParams
from purehtml.configs.types.config_with_selector import ConfigWithSelector


class ObjectConfig(ConfigWithSelector):
    """
    Concrete class representing an object configuration with selector and properties.
    """

    def __init__(self, selector: Optional[str], properties: Dict[str, Config]):
        """
        Initializes the ObjectConfig with a selector and properties map.
        :param selector: The selector string.
        :param properties: A dictionary mapping property names to Config objects.
        """
        super().__init__()
        self.selector = selector
        self.properties = properties

    def set_properties(self, properties: Dict[str, Config]):
        """
        Set the properties for this object configuration.
        :param properties: A dictionary mapping property names to Config objects.
        """
        self.properties = properties

    def extract(self, params: ExtractParams) -> Any:
        """
        Extract values for each property based on the provided parameters.
        :param params: The ExtractParams object containing context for extraction.
        :return: A dictionary of extracted property values.
        """

        parent = params.node()
        selector_params = GetSelectorMatchesParams(
            already_matched=params.get_element_already_matched() or False,
            include_root=False,
        )

        element = self.get_first_match(parent, selector_params, params.document())
        if not element and self.selector:
            return None

        props = self.properties
        result = {}

        for key in props:
            config = props[key]
            extract_params = ExtractParams(document=params.document(),
                                           nodes=[element],
                                           url=params.url(),
                                           element_already_matched=False)

            result[key] = config.extract(extract_params)

        return result

File: java/src/main/java/com/purescraps/purehtml/configs/types/ObjectConfig.java
package com.purescraps.purehtml.configs.types;

import java.util.*;

import com.purescraps.purehtml.ExtractParamsBuilder;
import com.purescraps.purehtml.backend.PureHTMLNode;
import com.purescraps.purehtml.configs.Config;
import com.purescraps.purehtml.interfaces.ExtractParams;
import com.purescraps.purehtml.interfaces.GetSelectorMatchesParams;

public class ObjectConfig extends ConfigWithSelector {
    // Properties map: key-value pairs of property names and Config objects
    private Map<String, Config> properties = new HashMap<>();

    public ObjectConfig(String selector, Map<String, Config> properties) {
        super();
        this.selector = selector;
        this.setProperties(properties);
    }

    public void setProperties(Map<String, Config> properties) {
        this.properties = properties;
    }

    @Override
    public Object extract(ExtractParams params) {

        List<PureHTMLNode> parent = params.node();
        GetSelectorMatchesParams param = new GetSelectorMatchesParams() {
            @Override
            public boolean isAlreadyMatched() {
                return params.getElementAlreadyMatched();
            }

            @Override
            public boolean isIncludeRoot() {
                return false;
            }

        };

        PureHTMLNode element = this.getFirstMatch(parent, param, params.document());

        if ((element == null) && (this.selector != null))
            return null;

        Map<String, Config> props = this.properties;

        Map<String, Object> result = new HashMap<>();
        for (String key : props.keySet()) {
            Config config = props.get(key);

            ExtractParams extractParams = new ExtractParamsBuilder()
                    .setDocument(params.document())
                    .setNode(new ArrayList<>(Collections.singletonList(element)))
                    .setUrl(params.url())
                    .setElementAlreadyMatched(params.getElementAlreadyMatched())
                    .build();
            result.put(key, config.extract(extractParams));
        }
        return result;
    }
}
